pkg/virtctl/imageupload/imageupload.go

```diff
@@ -0,0 +1,304 @@
+	cmd.MarkFlagRequired("uploadproxy-url")
+	cmd.Flags().StringVar(&pvcName, "pvc-name", "", "The destination PVC.")
+	cmd.MarkFlagRequired("pvc-name")
+	cmd.Flags().StringVar(&pvcSize, "pvc-size", "", "The size of the PVC to create (ex. 10Gi, 500Mi).")
+	cmd.Flags().StringVar(&storageClass, "storage-class", "", "The storage class for the PVC.")
+	cmd.Flags().StringVar(&accessMode, "access-mode", accessMode, "The access mode for the PVC.")
+	cmd.Flags().StringVar(&imagePath, "image-path", "", "Path to the local VM image.")
+	cmd.MarkFlagRequired("image-path")
+	cmd.Flags().BoolVar(&noCreate, "no-create", noCreate, "Don't attempt to create a new PVC.")
+	cmd.SetUsageTemplate(templates.UsageTemplate())
+	return cmd
+}
+
+func usage() string {
+	usage := `  # Upload a local disk image to a newly created PersistentVolumeClaim:
+	virtctl image-upload --insecure --upload-proxy-url=https://cdi-uploadproxy.mycluster.com --pvc-name=upload-pvc --pvc-size=10Gi --image-path=/images/fedora28.qcow2`
+	return usage
+}
+
+type command struct {
+	clientConfig clientcmd.ClientConfig
+}
+
+func (c *command) run(cmd *cobra.Command, args []string) error {
+	file, err := os.Open(imagePath)
+	if err != nil {
+		return err
+	}
+	defer file.Close()
+
+	namespace, _, err := c.clientConfig.Namespace()
+	if err != nil {
+		return err
+	}
+
+	virtClient, err := kubecli.GetKubevirtClientFromClientConfig(c.clientConfig)
+	if err != nil {
+		return fmt.Errorf("Cannot obtain KubeVirt client: %v", err)
+	}
+
+	pvc, err := getUploadPVC(virtClient, namespace, pvcName, noCreate)
+	if err != nil {
+		return err
+	}
+
+	if pvc == nil {
+		pvc, err = createUploadPVC(virtClient, namespace, pvcName, pvcSize, storageClass, accessMode)
+		if err != nil {
+			return err
+		}
+		fmt.Printf("PVC %s/%s created\n", namespace, pvc.Name)
+	} else {
+		fmt.Printf("Using existing PVC %s/%s\n", namespace, pvc.Name)
+	}
+
+	err = waitUploadPodRunning(virtClient, namespace, pvcName, uploadPodWaitInterval, uploadPodWaitTimeout)
+	if err != nil {
+		return err
+	}
+
+	token, err := getUploadToken(virtClient.CdiClient(), namespace, pvcName)
+	if err != nil {
+		return err
+	}
+
+	fmt.Printf("Uploading data to %s\n", uploadProxyURL)
+
+	err = uploadData(uploadProxyURL, token, file, insecure)
+	if err != nil {
+		return err
+	}
+
+	fmt.Printf("Uploading %s completed successfully\n", imagePath)
+
+	return nil
+}
+
+func getHTTPClient(insecure bool) *http.Client {
+	client := &http.Client{}
+
+	if insecure {
+		client.Transport = &http.Transport{
+			TLSClientConfig: &tls.Config{InsecureSkipVerify: true},
+		}
+	}
+
+	return client
+}
+
+func uploadData(uploadProxyURL, token string, reader io.Reader, insecure bool) error {
+	url := uploadProxyURL + uploadProxyURI
+	req, _ := http.NewRequest("POST", url, reader)
+	client := httpClientCreatorFunc(insecure)
+
+	req.Header.Add("Authorization", "Bearer "+token)
+	req.Header.Add("Content-Type", "application/octet-stream")
+
+	resp, err := client.Do(req)
```

Comment:
for big files, it would be really nice to have a progress indicator here to know how long the upload is taking.

Reply:
That would be nice!  Will investigate!